runtime-sdk/src/modules/consensus/mod.rs

```diff
@@ -316,6 +330,45 @@ impl API for Module {
 
         Ok(scaled)
     }
+
+    fn height_for_epoch<C: Context>(ctx: &C, epoch: EpochTime) -> Result<u64, Error> {
+        static HEIGHT_CACHE: Lazy<Mutex<lru::LruCache<EpochTime, u64>>> =
+            Lazy::new(|| Mutex::new(lru::LruCache::new(NonZeroUsize::new(128).unwrap())));
+
+        // Check the cache first to avoid more expensive traversals.
+        let mut cache = HEIGHT_CACHE.lock().unwrap();
+        if let Some(height) = cache.get(&epoch) {
+            return Ok(*height);
+        }
+
+        // Resolve height for the given epoch.
+        let mut height = HEIGHT_LATEST;
```

Comment:
Minor, probably irrelevant? improvement: could start at the height of the smallest cached epoch that is higher than `epoch`.